# Examen Parcial

## Importaciones

In [13]:
# Para visualizar más de una salida de una celda
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"

## Preguntas

### 1. 

Recordemos que las tuplas son diferentes de las listas porque son inmutables: no se pueden modificar. Esto significa que no se pueden agregar o quitar elementos de las tuplas, y los elementos no se pueden reemplazar dentro de las tuplas. Sin embargo, puede concatenar 2 o más tuplas para formar una nueva tupla. Escriba una función en Python a la que se le pase una tupla y cualquier cantidad de elementos para adicionarle y devuelva la nueva tupla.

In [160]:
def tuple_add(*args): 
    return sum((x for x in args),())  

my_tuple = (1,2,3,4)
my_tuple2 = (1,2,3,4)
my_tuple3 = tuple_add(my_tuple,my_tuple2)
my_tuple3


(1, 2, 3, 4, 1, 2, 3, 4)

### 2.

¿Cuál es el resultado del siguiente código?

In [7]:
aTuple = (100, 200, 300, 400, 500)
aTuple[1] = 800
print(aTuple)

TypeError: 'tuple' object does not support item assignment

### 3. 

Selecciona la opción donde todo sea verdad sobre los sets en Python

+ (X) Los sets están desordenados, los sets no permiten duplicados, los sets están escritos entre llaves {}
+ ( ) Los sets tienen indizado, los sets son mutables

### 4.

In [8]:
set1 = {10, 20, 30, 40, 50}
set2 = {60, 70, 10, 30, 40, 80, 20, 50} 

print(set1.issubset(set2))
print(set2.issuperset(set1))

True
True


### 5. 

¿Cómo se podría eliminar 'baz' del siguiente diccionario?  
![Alt text](image-1.png)

In [9]:
dic = {'foo':100,'bar':200,'baz':300}
del dic['baz']
dic

{'foo': 100, 'bar': 200}

### 6.

En Python los diccionarios son inmutables: Falso

### 7. 

Escriba un código en Python, usando comprensión de listas, que le permita encontrar los números entre 1 y 1000 (inclusivos), que tengan al menos un dígito '3' en ellos.

In [44]:
digitos_3 = list(num for num in range(1,1001) if ('3' in str(num)))
print(digitos_3)


[3, 13, 23, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 43, 53, 63, 73, 83, 93, 103, 113, 123, 130, 131, 132, 133, 134, 135, 136, 137, 138, 139, 143, 153, 163, 173, 183, 193, 203, 213, 223, 230, 231, 232, 233, 234, 235, 236, 237, 238, 239, 243, 253, 263, 273, 283, 293, 300, 301, 302, 303, 304, 305, 306, 307, 308, 309, 310, 311, 312, 313, 314, 315, 316, 317, 318, 319, 320, 321, 322, 323, 324, 325, 326, 327, 328, 329, 330, 331, 332, 333, 334, 335, 336, 337, 338, 339, 340, 341, 342, 343, 344, 345, 346, 347, 348, 349, 350, 351, 352, 353, 354, 355, 356, 357, 358, 359, 360, 361, 362, 363, 364, 365, 366, 367, 368, 369, 370, 371, 372, 373, 374, 375, 376, 377, 378, 379, 380, 381, 382, 383, 384, 385, 386, 387, 388, 389, 390, 391, 392, 393, 394, 395, 396, 397, 398, 399, 403, 413, 423, 430, 431, 432, 433, 434, 435, 436, 437, 438, 439, 443, 453, 463, 473, 483, 493, 503, 513, 523, 530, 531, 532, 533, 534, 535, 536, 537, 538, 539, 543, 553, 563, 573, 583, 593, 603, 613, 623, 630, 631, 632, 633, 634, 635,

### 8. 

Escriba un código en Python, usando comprensión de listas, que le permita encontrar todos los números entre 1 y 1000 (inclusivos) que sean divisibles entre 7

In [46]:
div_7 = list(num for num in range(1,1001) if (num%7==0))
print(div_7)

[7, 14, 21, 28, 35, 42, 49, 56, 63, 70, 77, 84, 91, 98, 105, 112, 119, 126, 133, 140, 147, 154, 161, 168, 175, 182, 189, 196, 203, 210, 217, 224, 231, 238, 245, 252, 259, 266, 273, 280, 287, 294, 301, 308, 315, 322, 329, 336, 343, 350, 357, 364, 371, 378, 385, 392, 399, 406, 413, 420, 427, 434, 441, 448, 455, 462, 469, 476, 483, 490, 497, 504, 511, 518, 525, 532, 539, 546, 553, 560, 567, 574, 581, 588, 595, 602, 609, 616, 623, 630, 637, 644, 651, 658, 665, 672, 679, 686, 693, 700, 707, 714, 721, 728, 735, 742, 749, 756, 763, 770, 777, 784, 791, 798, 805, 812, 819, 826, 833, 840, 847, 854, 861, 868, 875, 882, 889, 896, 903, 910, 917, 924, 931, 938, 945, 952, 959, 966, 973, 980, 987, 994]


### 9.

Complete el siguiente código Python que calcula el cuadrado y el cubo todos los números de la siguiente lista de enteros usando funciones lambda.

El resultado (salida del código) debe verse así:

Lista original: [1,2,3,4,5,6,7,8,9,10]

Lista al cuadrado: [1,4,9, 16, 25, 36, 49, 64, 81, 100]

Lista al cubo: [1, 8, 27, 64, 125, 216, 343, 512, 729, 1000]

In [48]:
nums = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10] 
print("Lista original:") 
print(nums) 
print("\nLista al cuadrado:") 
square_nums = list(map(lambda x: x**2,nums))


print(square_nums) 
print("\nLista al cubo:") 
cube_nums = list(map(lambda x: x**3,nums))

 
print(cube_nums)

Lista original:
[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

Lista al cuadrado:
[1, 4, 9, 16, 25, 36, 49, 64, 81, 100]

Lista al cubo:
[1, 8, 27, 64, 125, 216, 343, 512, 729, 1000]


### 10.

Escribe un programa Python para filtrar una lista de enteros usando lambda

El resultado debe verse así, incluyendo el texto:

Lista original: [1,2,,3,4,5,6,7,8,9,10]

Números pares de la lista: [2,4,6,8,10]

Números impares de la lista: [1,3,5,7,9]

In [52]:
lista = [1,2,3,4,5,6,7,8,9,10]
print("Lista original: ", lista)

pares = list(filter(lambda x: x%2==0,nums))
print("Números pares de la lista: ", pares)

impares = list(filter(lambda x: x%2!=0,nums))
print("Números impares de la lista: ", impares)



Lista original:  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
Números pares de la lista:  [2, 4, 6, 8, 10]
Números impares de la lista:  [1, 3, 5, 7, 9]


### 11.

Dada la siguiente función generadora que produce una contraseña al azar, seleccione todas las respuestas válidas:  

+ print ("Password is ", randomPassword()) imprime el password

+ X El ciclo siguiente imprime el password  
for i in randomPassword():
    print(i)


+ X El password tiene 8 caracteres

+ Si eliminamos la sentencia  
     random.SystemRandom().shuffle(passwordList)  
ya no obtenemos el password

In [61]:
import random
import string
def randomPassword():
    randomSource = string.ascii_letters + string.digits + string.punctuation
    password = random.sample(randomSource, 4)
    password += random.sample(string.ascii_uppercase, 2)
    password += random.choice(string.digits)
    password += random.choice(string.punctuation)
    passwordList = list(password)
    random.SystemRandom().shuffle(passwordList)
    password = ''.join(passwordList)
    yield password

print ("Password is ", next(randomPassword()))


Password is  M;J3NQ8-


### 12.

Complete el código para invertir un arreglo (el primer elemento se convierte en el último) usando operador de corte (slicing).

La salida del código debe verse así:

Arreglo original:

[12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37]

Arreglo invertido:

[37 36 35 34 33 32 31 30 29 28 27 26 25 24 23 22 21 20 19 18 17 16 15 14 13 12]

In [62]:
import numpy as np 
x = np.arange(12, 38) 
print("Arreglo original:") 
print(x) 
print("Arreglo invertido:") 
x = x[::-1]
print(x)

Arreglo original:
[12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35
 36 37]
Arreglo invertido:
[37 36 35 34 33 32 31 30 29 28 27 26 25 24 23 22 21 20 19 18 17 16 15 14
 13 12]


### 13.

Utilizando 2 arreglos 2D, complete el código con una sentencia para imprimir un nuevo arreglo que conserve la forma 2D y la misma cantidad de filas.

La salida debe verse:  
[[ 1  2  3  7  8  9]  
 [ 4  5  6 10 11 12]]



In [ ]:
a1 = np.array([[1,2,3],
               [4,5,6]])
a2 = np.array([[7,8,9],
               [10,11,12]])

print(_________________________)

### 15.

El archivo cotizacion.csv contiene las cotizaciones de las empresas del IBEX35 con las siguientes columnas: 
+ nombre (nombre de la empresa)
+ Final (precio de la acción al cierre de bolsa)
+ Máximo (precio máximo de la acción durante la jornada)
+ Mínimo (precio mínimo de la acción durante la jornada)
+ volumen (Volumen al cierre de bolsa)
+ Efectivo (capitalización al cierre en miles de euros).  

Construir una función que construya un DataFrame a partir de un archivo con el formato anterior y devuelva otro DataFrame con el mínimo, el máximo y la media de cada columna.

In [167]:
import pandas as pd

def df_calc(file_name):
    # Creación
    df = pd.read_csv(file_name, delimiter=';')
    calc_df = pd.DataFrame()
    calc_df['column'] = df.columns
    c_min,c_max,c_mean= [],[],[]


    for col in df.columns:
        # Limpieza
        df[col] = df[col].replace(',','.',regex=True)
        for i in df[col]:
            if (i.count('.') > 1):
                i.replace('.','') 
                

       ##Cálculos
        c_min.append(df[col].min())
        c_max.append(df[col].max())

        try:
            c_mean.append(df[col].mean())
            print('numeric:',col)
        except:
            c_mean.append(np.NaN)
            print('error', col)
            
    
    calc_df['min'] = c_min
    calc_df['max'] = c_max
    calc_df['mean'] = c_mean

    calc_df

    return calc_df
    

In [168]:
arch = 'cotizacion.csv'
calc_df = df_calc(arch)
calc_df

error Nombre
error Final
error Máximo
error Mínimo
error Volumen
error Efectivo


Nombre   Final  Máximo  Mínimo     Volumen   Efectivo
0        ACCIONA   95.95   96.75    94.4      84.962    8166.11
1       ACERINOX    8668    8672    8468      885.99    7633.81
2            ACS   37.28   37.66   37.22     655.982   24517.29
3           AENA   167.1   167.5   166.1     133.738   22336.10
4        AMADEUS   69.96   70.64   69.56     640.476   44795.47
5    ARCELORMIT.   14152   14318   14.08     702.841    9989.65
6    BA.SABADELL  1.0165    1033  1.0165  12.358.512   12637.65
7   BA.SANTANDER    4.02  4.0675    4.01  36.129.692  145765.44
8         BANKIA    2249    2262    2224   3.024.454    6803.54
9      BANKINTER    6.56    6604    6522   1.375.976    9030.20
10          BBVA    5024    5.05    4969  26.447.612  132560.15
11     CAIXABANK    2763    2806    2762  13.031.097   36239.22
12       CELLNEX   30.17    30.7   30.17     857.237   25965.29
13  CIE AUTOMOT.   23.44   23.74    23.4      99.506    2343.09
14        ENAGAS    25.6   25.66   25.07      842.54   21524.77
15          ENCE    3882    3974    3846   2.099.698    8219.83
16        ENDESA   22.95   23.03   22.62   1.236.799   28233.46
17     FERROVIAL   21.61   21.72   21.38   2.065.931   44656.05
18  GRIFOLS CL.A    23.3    23.6   23.27     719.503   16787.55
19           IAG     5.3    5.46    5292   2.155.081   11561.95
20     IBERDROLA    8.45    8.45    8262  15.637.958  131426.05
21       INDITEX   24.93   25.19   24.87   2.600.144   65012.92
22       INDRA A    9.56    9685    9535      392.06    3755.57
23  INM.COLONIAL    9915    9915    9745     671.753    6640.38
24        MAPFRE    2654    2654    2.62   3.045.972    8063.13
25      MEDIASET    6446    6592    6446   2.647.557   17277.04
26  MELIA HOTELS    8.14    8.19    8.02      480.21    3914.93
27        MERLIN   12.08   12.08   11.83     777.543    9357.77
28       NATURGY    26.7   26.77   26.53   2.100.737   55986.54
29        R.E.C.   19705   19875   19675   1.246.171   24585.33
30        REPSOL   14515   14605   14.46   5.046.648   73333.78
31  SIEMENS GAME   14.33   14.39   14155   1.054.847   15084.56
32  TEC.REUNIDAS      23    23.2   22.78       122.1    2807.19
33    TELEFONICA    7204    7256    7185   9.913.249   71564.99
34      VISCOFAN   48.46   49.12   48.46       67.71    3295.43

In [144]:
df


Nombre   Final  Máximo  Mínimo     Volumen   Efectivo
0        ACCIONA   95.95   96,75    94,4      84.962    8166,11
1       ACERINOX    8668    8672    8468      885.99    7633,81
2            ACS   37.28   37,66   37,22     655.982   24517,29
3           AENA   167.1   167,5   166,1     133.738   22336,10
4        AMADEUS   69.96   70,64   69,56     640.476   44795,47
5    ARCELORMIT.   14152   14318   14,08     702.841    9989,65
6    BA.SABADELL  1.0165    1033  1,0165  12.358.512   12637,65
7   BA.SANTANDER    4.02  4,0675    4,01  36.129.692  145765,44
8         BANKIA    2249    2262    2224   3.024.454    6803,54
9      BANKINTER    6.56    6604    6522   1.375.976    9030,20
10          BBVA    5024    5,05    4969  26.447.612  132560,15
11     CAIXABANK    2763    2806    2762  13.031.097   36239,22
12       CELLNEX   30.17    30,7   30,17     857.237   25965,29
13  CIE AUTOMOT.   23.44   23,74    23,4      99.506    2343,09
14        ENAGAS    25.6   25,66   25,07      842.54   21524,77
15          ENCE    3882    3974    3846   2.099.698    8219,83
16        ENDESA   22.95   23,03   22,62   1.236.799   28233,46
17     FERROVIAL   21.61   21,72   21,38   2.065.931   44656,05
18  GRIFOLS CL.A    23.3    23,6   23,27     719.503   16787,55
19           IAG     5.3    5,46    5292   2.155.081   11561,95
20     IBERDROLA    8.45    8,45    8262  15.637.958  131426,05
21       INDITEX   24.93   25,19   24,87   2.600.144   65012,92
22       INDRA A    9.56    9685    9535      392.06    3755,57
23  INM.COLONIAL    9915    9915    9745     671.753    6640,38
24        MAPFRE    2654    2654    2,62   3.045.972    8063,13
25      MEDIASET    6446    6592    6446   2.647.557   17277,04
26  MELIA HOTELS    8.14    8,19    8,02      480.21    3914,93
27        MERLIN   12.08   12,08   11,83     777.543    9357,77
28       NATURGY    26.7   26,77   26,53   2.100.737   55986,54
29        R.E.C.   19705   19875   19675   1.246.171   24585,33
30        REPSOL   14515   14605   14,46   5.046.648   73333,78
31  SIEMENS GAME   14.33   14,39   14155   1.054.847   15084,56
32  TEC.REUNIDAS      23    23,2   22,78       122.1    2807,19
33    TELEFONICA    7204    7256    7185   9.913.249   71564,99
34      VISCOFAN   48.46   49,12   48,46       67.71    3295,43